Propertieslayout Fields And Forms Redesign

signedav opened the pull request:
## Description
Source Fields and Attribute Form configuration are splitted.
Would be nice to have this included in QGIS 3.0 to give it a fresh new UX.

### Source Fields 
![sourcefields](https://user-images.githubusercontent.com/28384354/32061615-6a60088a-ba72-11e7-86b8-3a6e5b1a1fd9.png)
Yellow are the source fields, virtual fields are blue and green are the joined fields.

### Todos Source Fields
- [ ] Think about where to put the WFS stuff (optional) 
- [ ] Move button of virtual fields to start of line (optional) 

### Attribute Form
![layerproperties](https://user-images.githubusercontent.com/28384354/32061647-7cddca6a-ba72-11e7-8a39-2a98386e25bc.png)

### Todos Attribute Form
- [x] Store the Alias and the Comment
- [x] Configuration for relations (f.e. cardinality)
- [x] Show label field
- [x] Button to python init functions
- [ ] Preview option (optional) 
- [ ] Improve layout of widgets in Drag-List (optional)
- [x] Fix trouble with first feature

Todos marked as optional don't have to be done for 3.0...

nyalldawson committed:
Add items for project map layers and relations to expression builder

Allows easy insertion of map layer IDs and relation IDs into
expressions. Numerous expression functions now utilise these,
so it makes sense to allow them to be easily inserted.

Fix #11680, #16879

Sponsored by Andreas Neumann

signedav committed:
QEP 37: Fields And Forms Redesign - interim status

signedav committed:
work in progress status

m-kuhn committed:
Code formatting

m-kuhn committed:
Add getter/setter for type

m-kuhn committed:
m prefix is only for members

m-kuhn committed:
Range-based for loop

m-kuhn committed:
Remove switch default: Force warnings

m-kuhn committed:
Descriptive Titles

m-kuhn committed:
Change widget selection list to combobox

m-kuhn committed:
Merge remote-tracking branch 'qgis/master' into propertieslayout

m-kuhn committed:
Fix build warnings

m-kuhn committed:
Deal with QGIS' need to talk

m-kuhn committed:
Disable QgsFieldsProperites save logic

m-kuhn committed:
Cleanup slots

m-kuhn committed:
Avoid warnings

m-kuhn committed:
Fix reference to temporary

m-kuhn committed:
Range based for loop

m-kuhn committed:
Silence of the lambs

m-kuhn committed:
Fix store/reload of widget configuration

m-kuhn committed:
Some layout changes

m-kuhn committed:
Remove debug statement

signedav committed:
Merge pull request #1 from m-kuhn/propertieslayout

Propertieslayout

signedav committed:
add relations into drag-tree

signedav committed:
relations in list
alias comment in gui (not ready yet)

signedav committed:
Merge branch 'master' into propertieslayout

signedav committed:
Merge branch 'master' into propertieslayout

signedav committed:
alias field and fixed problem with storing

signedav committed:
relationconfig

signedav committed:
Relation Cardinality without own gui

signedav committed:
relation properties

signedav committed:
Relation Cardinality config

signedav committed:
combobox for cardinality

signedav committed:
commit

signedav committed:
commit

signedav committed:
Relation Combo

signedav committed:
spelling

signedav committed:
style

signedav committed:
layout

signedav committed:
App doc and style changes

signedav committed:
final relation cardinality combobox

signedav committed:
attribute editor layout selection etc.

signedav committed:
basics of new dialog for the initcode python functionalities
means, some cosmetic changes should still be done in qgsattributesforminitcode.ui

m-kuhn commented on src/app/qgsattributesforminitcode.cpp:
```diff
@@ -0,0 +1,84 @@
+#include "qgsattributesforminitcode.h"
```
License headers

m-kuhn commented on src/app/qgsattributesforminitcode.h:
```diff
@@ -0,0 +1,70 @@
+
+#ifndef QGSATTRIBUTESFORMINITCODE_H
+#define QGSATTRIBUTESFORMINITCODE_H
+/***************************************************************************
```
Extra license header here ;)

ninsbl commented:
As for alias and comments. Is their storage coordinated with: https://github.com/qgis/QGIS-Enhancement-Proposals/issues/91
I would assume that both are vital aspects of "Metadata", and at least fields/attributes are covered by ISO, right?
Or is it rather the other way around, that the metadata work can/should pickup data stored for the fields other places...

m-kuhn commented:
> Or is it rather the other way around, that the metadata work can/should pickup data stored for the fields other places...

That's what I would suppose. Keeping this together as "form configuration" makes sense to me in scenarios without any metadata requirements involved. Please raise this question on QEP#91 if you think the metadata aspect of these configuration options need special attention.

DelazJ commented:
Good to see this pr land and would be a nice change for 3.0
I thought that in the qep, it was agreed that widget and form layout frames would be [vertically aligned](https://github.com/qgis/QGIS-Enhancement-Proposals/issues/37#issuecomment-316627554)?These two blank spaces at the bottom are not that nice imho. Also, it's not obvious that the third part are options from the widget because of the separation (by another frame with selectable item). 

Also something that has puzzled me in this dialog and that was somehow fixed in the [recent refactoring](https://github.com/qgis/QGIS/pull/3771) of this dialog is that some of the options displayed at right (editable checkbox, label on top, constraints, maybe default) are related to the field and not to the selected widget type and this difference is not clear with the way they are regrouped now imho.

andreasneumann commented:
I think it would be good if in the "Source Fields" tab the "E" (Expression button) would be moved from the "Comment" column to the very front, where we already have an icon indicating that it is an expression field. Replacing this icon with a button with Icon would be good - as you already suggested as an option.

andreasneumann commented:
Other question: where are the Python init functions specified and where can I select the form type (Autogenerate, Drag and Drop and UI-File)?

Or will the Autogenerate type be replaced by a standard drag and drop form where everything is in one tab and column?

m-kuhn commented:
Thanks for the comments @DelazJ 

@signedav is out of office today, so I'll try to respond for him (@signedav, correct me where wrong, please)

IIRC @nirvn had some good comments on vertical alignment.

Strange, that alignment fixes have gone lost, they have been in there a long time before branching for this pull request.

@andreasneumann 

Looking at the more recent commit messages, layout and init functions have been integrated after this pr has been opened.

I agree for the expression button, but would propose to make this an optional extra which can also be pushed for 3.2 and shouldn't block the rest.

andreasneumann commented:
The two ungrouped checkboxes "Editable" and "Label on top" could go in the "General" group on the top and/or in the "Widget Type" group.

andreasneumann commented:
@m-kuhn @signedav - it is fine to postpone the moving of the "E" to 3.2 for me - if it is too complicated to do. If it simple, I would still do it now.

signedav commented:
Good morning. Sorry, I was unavailable this weekend.

You are right @DelazJ that we tended to do the vertical alignment in the [discussion](https://github.com/qgis/QGIS-Enhancement-Proposals/issues/37#issuecomment-316627554).
Why I did it the other way anyway, was because we had some discussion at the last hackfest and a lot of voices for the other design (the one with separate configuration of the widged and the layout like drag&drop). So we decided for a compromise solution, like the one implemented. I admit, I should have had summarize it to the online discussion too. The reason for the horizontal alignment is, that many people often use a lot of attributes.

@andreasneumann Yes, I remember your input with the expression button. I'll see if it's easy to do. The python init function and the form type are now included. At the moment of the first screenshot I didn't had it yet on the GUI. 
![image](https://user-images.githubusercontent.com/28384354/32428265-2b4882de-c2c5-11e7-8592-bc3970cb1c56.png)

With seperate dialog for the python stuff:
![image](https://user-images.githubusercontent.com/28384354/32428291-501b413c-c2c5-11e7-899f-d05e42778aef.png)

m-kuhn commented on src/app/qgsattributerelationedit.cpp:
```diff
@@ -0,0 +1,31 @@
+#include "qgsattributerelationedit.h"
```
License header

m-kuhn replied:
Options -> C++ -> File Naming -> License Template -> `[prefix]/QGIS/editors/QtCreator/qt_creator_license_template`

m-kuhn commented on src/app/qgsattributerelationedit.h:
```diff
@@ -0,0 +1,66 @@
+
+class QgsAttributeRelationEdit : public QWidget
+{
+*/
```
Required?

m-kuhn commented on src/app/qgsattributesforminitcode.cpp:
```diff
@@ -0,0 +1,84 @@
+  setupUi( this );
+
+  // Init function stuff
+  mInitCodeSourceComboBox->addItem( tr( "" ) );
```
s/tr( "" )/QString()/

tr() is for translation, I guess the empty string is the same for all locales

m-kuhn commented on src/app/qgsattributesforminitcode.cpp:
```diff
@@ -0,0 +1,84 @@
+
+QgsEditFormConfig::PythonInitCodeSource QgsAttributesFormInitCode::codeSource() const
+{
+  return ( QgsEditFormConfig::PythonInitCodeSource )mInitCodeSourceComboBox->currentIndex();
```
Maybe use currentData here

    return mInitCodeSourceComboBox->currentData()

and in the initialization

    addItem->( text, sourceType )

m-kuhn commented on src/app/qgsattributesformproperties.cpp:
```diff
@@ -0,0 +1,1174 @@
+#include "qgsattributesformproperties.h"
```
License header

m-kuhn commented on src/app/qgsattributesformproperties.cpp:
```diff
@@ -0,0 +1,1174 @@
+
+    QTreeWidgetItem *item = mAvailableWidgetsTree->addItem( catitem, itemData );
+    item->setData( 0, RelationConfigRole, cfg );
+    item->setData( 0, FieldNameRole, QStringLiteral( "%1" ).arg( relation.id() ) ); //relation.name() );
```
No reason for the `%1` I guess, just use the id directly

m-kuhn commented on src/ui/qgsattributesforminitcode.ui:
```diff
@@ -0,0 +1,210 @@
+
+An example is:
+
+          from PyQt4.QtGui import QWidget
```
from qgis.PyQt.QtWidgets import QWidget

m-kuhn commented on src/app/qgsattributetypedialog.h:
```diff
@@ -63,6 +64,26 @@ class APP_EXPORT QgsAttributeTypeDialog: public QDialog, private Ui::QgsAttribut
+    /**
+     * Getter for lable comment
+     */
+    QString comment();
```
Const

m-kuhn commented on src/app/qgsattributetypedialog.h:
```diff
@@ -63,6 +64,26 @@ class APP_EXPORT QgsAttributeTypeDialog: public QDialog, private Ui::QgsAttribut
+    /**
+     * Getter for lable alias
+     */
+    QString alias();
```
Const,

Sidenote: method documentation should normally provide more information than implicitly available from the method name (i.e. do not state the obvious). It's not critical in here since it's in app, so no public api.

m-kuhn commented on src/app/qgsattributetypedialog.cpp:
```diff
@@ -328,18 +336,39 @@ void QgsAttributeTypeDialog::setFieldEditable( bool editable )
+//removed selectionListWidget:   const QString editType = selectionListWidget->item( index )->data( Qt::UserRole ).toString();
+//removed selectionListWidget:
+//removed selectionListWidget:   setEditorWidgetType( editType );
+//removed selectionListWidget: }
```
Is that still relevant?

m-kuhn commented on src/app/qgsattributesformproperties.h:
```diff
@@ -0,0 +1,291 @@
+
+    QTreeWidgetItem *loadAttributeEditorTreeItem( QgsAttributeEditorElement *const widgetDef, QTreeWidgetItem *parent, DnDTree *tree );
+
+  protected:
```
I guess this will never be subclassed: make this private

m-kuhn replied:
Bump

m-kuhn commented:
Code looks good to me for the biggest part, good job so far!
The biggest challenge is probably not on the code though but on usability and layout, which is hard to review directly in code on github and will need to be tested thoroughly by just using it.

DelazJ commented on src/app/qgsattributetypedialog.cpp:
```diff
@@ -93,7 +89,11 @@ QgsAttributeTypeDialog::QgsAttributeTypeDialog( QgsVectorLayer *vl, int fieldIdx
 
   constraintExpressionWidget->setLayer( vl );
+
+  // TODO, what should this help be connected to?
```
I'd say: to somewhere in this section http://docs.qgis.org/testing/en/docs/user_manual/working_with_vector/vector_properties.html#fields-properties. Isn't it [already implemented ](https://github.com/qgis/QGIS/pull/5467/files#diff-be501c6a669cec8cc989f027704b1e2eR423) (unless I misunderstood - which is really probable)?

signedav replied:
Solved it now like this, that if you are on this tab (Form) this help-site is loaded when pressing help button of vectorlayerproperties dialog. Because it does not make sence to have two help-buttons.

m-kuhn commented on src/app/qgsattributetypedialog.h:
```diff
@@ -63,6 +64,26 @@ class APP_EXPORT QgsAttributeTypeDialog: public QDialog, private Ui::QgsAttribut
     bool labelOnTop() const;
 
     /**
+     * Setter for lable alias
```
Actually it should be field, not label ;)

signedav replied:
Yes, sorry guys. I guess lineedit is the right one.

DelazJ commented on src/app/qgsvectorlayerproperties.cpp:
```diff
@@ -86,6 +88,15 @@ QgsVectorLayerProperties::QgsVectorLayerProperties(
+  , diagramPropertiesDialog( nullptr )
+  , mFieldsPropertiesDialog( nullptr )
+  , mSourceFieldsPropertiesDialog( nullptr )
+  , mAttributesFormPropertiesDialog( nullptr )
```
@m-kuhn could this initialisation be moved to the .h? just /me curious

m-kuhn replied:
Yep

    s/could/should

signedav committed:
some changes in attribute form properties and source data properties
changed style of layout, initcode and suppress-combo
display read only comment
fixed saving of wfs
improved layout
some tooltips

signedav committed:
fixed problem with saving first entry

signedav committed:
help and suppress form
help from attributetypedialog is handled over the help of vectorlayerproperties
suppress form combo takes the default value from the settings
fixed troubles on loading gui

signedav committed:
layout and remove fieldproperties
icons and made naming
inputs from the pull request
remove old fieldpropertie-stuff

signedav committed:
item for source field fix

signedav commented:
Choosed these icons for the two property forms:

![image](https://user-images.githubusercontent.com/28384354/32457017-d8343b70-c327-11e7-8590-2f6bc8e16634.png)

signedav committed:
editable and lable on top + fixes
in attribute form properties: editable and lable on top config to general
fix: in case that somebody deletes a field on sourcedata and then wants to change config on it in attribute - it's now disabled (but not updated)
fix: if adding new attribute after adding a virtual one, the display of the list was wrong (value of virtual in non-virtual row) - now it reloads the table correctly

m-kuhn commented on src/app/qgssourcefieldsproperties.cpp:
```diff
@@ -281,6 +281,7 @@ bool QgsSourceFieldsProperties::addAttribute( const QgsField &field )
 void QgsSourceFieldsProperties::apply()
 {
   // Disabled: we deal with the configuration in the new tabs
```
Is this comment still valid?

m-kuhn commented on src/app/qgsattributesformproperties.cpp:
```diff
@@ -656,15 +647,15 @@ void QgsAttributesFormProperties::apply()
-    QString name = mLayer->fields().at( idx ).name();
-    FieldConfig cfg = configForChild( idx );
+    //continue in case field does not exist anymore
+    if ( idx < 0 ) continue;
```
Can you add a break before the continue? Just for the part of the code style that astyle still doesn't do ;)

m-kuhn commented on src/app/qgsattributesformproperties.cpp:
```diff
@@ -646,6 +625,18 @@ void QgsAttributesFormProperties::pbnSelectEditForm_clicked()
 
+int QgsAttributesFormProperties::getFieldIndexByName( const QString &name )
+{
+  for ( int idx = 0; idx < mLayer->fields().size(); idx++ )
```
mLayer.fields().indexOf( name )

signedav committed:
References #17412 Fixes problem when changeing sort but not when renaming
When sort and then toggle edditing, the problem was that it wanted to rename all the fields
Whis is avoided now by checking the indexes
But still problems open when one wants to rename fields in unsorted layout

signedav committed:
Fix #17412 Sorting fixed, no troubles when editing after sorting
Issue was, that the cellchange is triggered at re-sorting - so we need to check the names
Following issue was, that when removing or renaming something in re-sorted table - so there was a bug that referenced to the row instead of the index
Thought about to remove mIndexedWidgets completely, but when renumbering after delete in unsorted table, we cannot reference to the row order.

signedav committed:
no +/- buttons when not needed and some graphical finetuning

scroll area for attribute type config, but no scroll area for whole tab
set attribute type config to invisible when a container box is selected

NathanW2 commented:
Woot! Great work!

On Thu, Nov 9, 2017 at 2:46 AM, Matthias Kuhn <notifications@github.com>
wrote:

> Merged #5467 <https://github.com/qgis/QGIS/pull/5467>.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/5467#event-1332249718>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AAXS3G4YKN0-jRK_RQDMTJtwxsL-eBvtks5s0drWgaJpZM4QHxAT>
> .
>

DelazJ commented:
Just to report that without the [Feature] tag, I'm not sure this pull-request has been reported in the changelog. If someone can handle that...

m-kuhn commented:
Thanks @DelazJ .
Even with the tag, I'm not sure the current changelog has been created based on the commit messages, would be nice to get a feedback on this @timlinux @Gustry .
